src/test/java/io/confluent/examples/streams/AggregateTest.java

```diff
@@ -0,0 +1,116 @@
+import org.apache.kafka.common.serialization.StringDeserializer;
+import org.apache.kafka.common.serialization.StringSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.StreamsBuilder;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.TopologyTestDriver;
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.kstream.Grouped;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Materialized;
+import org.apache.kafka.streams.kstream.Produced;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * How to aggregate messages via `groupBy()` and `aggregate()`.
+ *
+ * This example can be adapted to structured (nested) data formats such as Avro or JSON in case you need to aggregate
+ * only certain field(s) in the input.
+ *
+ * Note: This example uses lambda expressions and thus works with Java 8+ only.
+ *
+ * See {@link AggregateScalaTest} for the equivalent Scala example.
+ */
+public class AggregateTest {
+
+  private static final String inputTopic = "inputTopic";
+  private static final String outputTopic = "outputTopic";
+
+  @Test
+  public void shouldAggregate() {
+    final List<String> inputValues = Arrays.asList(
+        "stream", "all", "the", "things", "hi", "world", "kafka", "streams", "streaming"
+    );
+
+    final Map<String, Long> expectedOutput = new HashMap<>();
+    expectedOutput.put("a", 3L);
+    expectedOutput.put("t", 9L);
+    expectedOutput.put("h", 2L);
+    expectedOutput.put("w", 5L);
+    expectedOutput.put("k", 5L);
+    expectedOutput.put("s", 22L);
+
+    // Step 1: Create the topology and its configuration
+    final StreamsBuilder builder = createTopology();
+    final Properties streamsConfiguration = createTopologyConfiguration();
+
+    try (final TopologyTestDriver testDriver = new TopologyTestDriver(builder.build(), streamsConfiguration)) {
+      // Step 2: Write the input
+      IntegrationTestUtils.produceValuesSynchronously(inputTopic, inputValues, testDriver, new StringSerializer());
+
+      // Step 3: Validate the output
+      final Map<String, Long> actualOutput = IntegrationTestUtils.drainTableOutput(
+          outputTopic, testDriver, new StringDeserializer(), new LongDeserializer());
+      assertThat(actualOutput).hasSameSizeAs(expectedOutput);
+      assertThat(actualOutput).containsAllEntriesOf(expectedOutput);
+    }
+  }
+
+  private StreamsBuilder createTopology() {
+    final StreamsBuilder builder = new StreamsBuilder();
+    final KStream<byte[], String> input = builder.stream(inputTopic, Consumed.with(Serdes.ByteArray(), Serdes.String()));
```

Comment:
Hmm, what I don't like about `NothingSerde<T>` is that it is not a default KStreams serde.  I think it is ok to use it when writing test data (via `IntegrationTestUtils`) and reading test output (via `IntegrationTestUtils`), but I am not convinced we should use it for the "real" example code. (For instance, you cannot easily copy-paste such snippets into your own code, or into a presentation slide, because outside of `kafka-streams-examples` the code does not work.)

Reply:
This is a very good point... I'll make a ticket to add the NothingSerde to Serdes.

Reply:
shaboom: https://issues.apache.org/jira/browse/KAFKA-8455

Thanks for the feedback, I agree 100%